### ISSUE TITLE
CORS policy for API Gateway issue

### ISSUE THREAD

**Issue by erashdan** [aws/serverless-application-model#1983]
Hello,

### Description:
I'm facing CORS policy issues only for endpoints that required authorization.

### Steps to reproduce:
- I created `AWS::Serverless::Api` in my template
```
RestApi:
    Type: AWS::Serverless::Api
    Properties:
      StageName: Prod
      Cors: "'*'"
      Auth:
        DefaultAuthorizer: MyCognitoAuthorizer
        AddDefaultAuthorizerToCorsPreflight: false
        ApiKeyRequired: false
        Authorizers:
          MyCognitoAuthorizer:
            UserPoolArn: !GetAtt UserPool.Arn
```
- After that, I created a function and linked it with this API
```
  Type: AWS::Serverless::Function
    Properties:
      Handler: src/handlers/get-a-test.handler
// etc
      Events:
        GetRoot:
          Type: Api
          Properties:
            RestApiId: !Ref RestApi
            Path: /test2
```

### Observed result:
Call API Gateway from web browsers only for unauthorized users
Access to XMLHttpRequest at 'https://XYZ.execute-api.us-east-1.amazonaws.com/Prod/test2' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

### Expected result:
To return 401 

### SAM Version

SAM CLI, version 1.22.0

Thanks

**Comment by brewsfab** [comment]
Hello @erashdan 

I would like to provide some insights.

TLDR:
Please investigate the Lambda function.



I do not have much details about your application but for having tested CORS in SAM locally, I realized that the error message is somehow misleading even if it's correct.

In fact, you got 

Access to XMLHttpRequest at 'https://XYZ.execute-api.us-east-1.amazonaws.com/Prod/test2' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

But it's might not be related to API Gateway directly but to the Lambda function.

Imagine that you handler is like this

```python

import json

def lambda_handler(event, context):

   # Your logic


    # If everything went well up to here
    return {
        'statusCode': 401,
        'headers': {
            'Access-Control-Allow-Headers': 'Content-Type',
            'Access-Control-Allow-Origin': 'https://www.example.com',
            'Access-Control-Allow-Methods': 'OPTIONS,POST,GET'
        },
        'body': json.dumps('Hello from Lambda!')
    }

```


In that case if you have a bug, an exception before reaching the last **return**, Lambda will send the error back to API Gateway without the specific headers and you will end up with a CORS issue.

If you could test locally with 
```
sam local start-api --debug
```

You will get something like that

```
/OPTIONS /test2 200
/GET /test2 xxx
```
xxx as 502 or any other specific error code.

Meaning the preflight worked well (200) but the Lambda had an issue.

Please investigate your Lambda.

Hope it helps.

**Comment by erashdan** [comment]
@brewsfab
- I removed Auth section from template and deploy it again
- Try to call endpoint /test2 
- It's work, so I think the issue `AddDefaultAuthorizerToCorsPreflight: False` to working

**Comment by sriram-mv** [comment]
@erashdan how does your options property within Open API section look like with and without `AddDefaultAuthorizerToCorsPreflight` being set.

**Comment by erashdan** [comment]
@sriram-mv it exported this 
```
openapi: "3.0.1"
info:
  title: "test-retro-auth"
  version: "1.0"
servers:
- url: "https://i77d48hrua.execute-api.us-east-1.amazonaws.com/{basePath}"
  variables:
    basePath:
      default: "/Prod"
paths:
  /test:
    get:
      security:
      - MyCognitoUserPoolAuthorizer: []
    options:
      responses:
        "200":
          description: "200 response"
          headers:
            Access-Control-Allow-Origin:
              schema:
                type: "string"
            Access-Control-Allow-Methods:
              schema:
                type: "string"
            Access-Control-Allow-Headers:
              schema:
                type: "string"
          content: {}
components:
  securitySchemes:
    MyCognitoUserPoolAuthorizer:
      type: "apiKey"
      name: "Authorization"
      in: "header"
      x-amazon-apigateway-authtype: "cognito_user_pools"

```

**Comment by erashdan** [comment]
@sriram-mv it exported this
```
openapi: "3.0.1"
info:
  title: "test-auth"
  version: "1.0"
servers:
- url: "https://i77d48hrua.execute-api.us-east-1.amazonaws.com/{basePath}"
  variables:
    basePath:
      default: "/Prod"
paths:
  /test:
    get:
      security:
      - MyCognitoUserPoolAuthorizer: []
    options:
      responses:
        "200":
          description: "200 response"
          headers:
            Access-Control-Allow-Origin:
              schema:
                type: "string"
            Access-Control-Allow-Methods:
              schema:
                type: "string"
            Access-Control-Allow-Headers:
              schema:
                type: "string"
          content: {}
components:
  securitySchemes:
    MyCognitoUserPoolAuthorizer:
      type: "apiKey"
      name: "Authorization"
      in: "header"
      x-amazon-apigateway-authtype: "cognito_user_pools"
```

**Comment by erashdan** [comment]
I solved this issue by add Access-Control-Allow-Origin to Gateway response
```
MyApi:
    Type: AWS::Serverless::Api
    Properties:
      StageName: Prod
      Cors: "'*'"
      GatewayResponses:
        UNAUTHORIZED:
          StatusCode: 401
          ResponseParameters:
            Headers:
              Access-Control-Allow-Origin: "'*'"
      Auth:
        DefaultAuthorizer: MyCognitoUserPoolAuthorizer
        AddDefaultAuthorizerToCorsPreflight: False
        Authorizers:
          MyCognitoUserPoolAuthorizer:
            UserPoolArn: !GetAtt MyCognitoUserPool.Arn

```
